src/uct/ib/mlx5/ib_mlx5.inl

```diff
@@ -72,28 +84,155 @@ uct_ib_mlx5_update_db_cq_ci(uct_ib_mlx5_cq_t *cq)
 }
 
 
+/**
+ * Fill the given CQE by unique data from given mini_cqe.
+ *
+ * @param cqe       The destination CQE.
+ * @param mini_cqe  The source mini CQE.
+ * @param cq_decomp The structure that contains info about CQE compression block.
+ */
+static UCS_F_ALWAYS_INLINE void
+uct_ib_mlx5_iface_cqe_decompress_fill_unique_fields(
+    struct mlx5_cqe64 *cqe, uct_ib_mlx5_mini_cqe8_t *mini_cqe,
+    uct_ib_mlx5_cq_decompress_t *cq_decompress)
+{
+    uint32_t host_sop_drop_qpn;
+    uint8_t  host_s_wqe_opcode;
+
+    /* TODO: Think about the checksum field decompression.
+     * The checksum field does not copied from src dst
+     * because of absence of the mlx5_cqe64::check_sum field on some systems.
+     */
+    cqe->byte_cnt           = mini_cqe->byte_cnt;
+    ucs_assert(!(cqe->op_own & UCT_IB_MLX5_CQE_OP_OWN_ERR_MASK));
+    if ((cqe->op_own >> 4) == MLX5_CQE_REQ) { /* Requester */
+        cqe->wqe_counter    = mini_cqe->wqe_counter;
+
+        /* Set s_wqe_opcode */
+        host_sop_drop_qpn   = ntohl(cqe->sop_drop_qpn);
+        host_s_wqe_opcode   = ntohs(mini_cqe->s_wqe_opcode);
+        host_sop_drop_qpn  &= UCS_MASK(UCT_IB_QPN_ORDER);
+        host_sop_drop_qpn  |= host_s_wqe_opcode << UCT_IB_QPN_ORDER;
+        cqe->sop_drop_qpn   = htonl(host_sop_drop_qpn);
+    } else { /* Responder */
+        cqe->wqe_counter    = htons(cq_decompress->wqe_counter +
+                                    cq_decompress->block_size -
+                                    cq_decompress->left);
+    }
+}
+
+
+/**
+ * Continue decompressing the CQEs. Should be used only when cq_decompress->left > 0.
+ *
+ * @param cq CQ that contains unhandled compressed CQEs (cq_decompress->left > 0).
+ *
+ * @return Next decompressed CQE.
+ */
 static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
-uct_ib_mlx5_poll_cq(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq)
+uct_ib_mlx5_iface_cqe_decompress_continue(uct_ib_mlx5_cq_t *cq)
 {
+    uct_ib_mlx5_cq_decompress_t *cq_decompress
+                                      = &cq->cq_decompress;
+    uint8_t id                        = cq_decompress->mini_arr_idx;
+    struct mlx5_cqe64 *title          = &cq_decompress->title;
+    uct_ib_mlx5_mini_cqe8_t *mini_cqe = &cq_decompress->mini_arr[id % 8];
     struct mlx5_cqe64 *cqe;
-    unsigned cqe_index;
-    uint8_t op_own;
 
-    cqe_index = cq->cq_ci;
-    cqe       = uct_ib_mlx5_get_cqe(cq, cqe_index);
-    op_own    = cqe->op_own;
+    uct_ib_mlx5_iface_cqe_decompress_fill_unique_fields(title, mini_cqe,
+                                                        cq_decompress);
 
-    if (ucs_unlikely(uct_ib_mlx5_cqe_is_hw_owned(op_own, cqe_index, cq->cq_length))) {
-        return NULL;
-    } else if (ucs_unlikely(op_own & UCT_IB_MLX5_CQE_OP_OWN_ERR_MASK)) {
+    cq_decompress->left--;
+    cq_decompress->mini_arr_idx++;
+    cq->cq_ci++;
+
+    if (cq_decompress->left != 0) {
+        if (cq_decompress->mini_arr_idx % 8 == 0) {
```

Comment:
1. can we define the "8" constant ? it's used multiple times
2. add ( ) around % expression